### PR TITLE
LA-293 Login with second factor authentication

### PR DESCRIPTION
**Pull request by manhduc-nguyen**
https://user-images.githubusercontent.com/16035728/113555747-9e604e00-9625-11eb-9ba9-f46cdb9d1d05.mp4

### CODE REVIEW

**Comment by hoangdat** [comment]
quite good!. Go ahead!!!

File: data/lib/src/datasource/authentication_datasource.dart [hunk 1]
```diff
@@ -45,18 +45,22 @@ class AuthenticationDataSource {
...
           userName.userName,
           password.value,
-          PermanentTokenBodyRequest('LinShare-${deviceManager.getPlatformString()}-$deviceUUID'));
+          PermanentTokenBodyRequest('LinShare-${deviceManager.getPlatformString()}-$deviceUUID'),
+          otpCode: otpCode);
       return permanentToken.toToken();
     }).catchError((error) {
         _remoteExceptionThrower.throwRemoteException(error, handler: (DioError error) {
           if (error.response.statusCode == 401) {
+            if (isAuthenticateWithOTPError(error.response.headers.value('X-Linshare-Auth-Error-Code'))) {
```
Comment (hoangdat) [inline data/lib/src/datasource/authentication_datasource.dart:61]: `X-Linshare-Auth-Error-Code` why not constant it?

File: ios/Runner/Info.plist [hunk 2]
```diff
@@ -2,6 +2,11 @@
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
+    <key>LSApplicationQueriesSchemes</key>
+    <array>
+      <string>https</string>
+      <string>http</string>
+    </array>
```
Comment (hoangdat) [inline ios/Runner/Info.plist:9]: can you say sth about that?

Comment (manhduc-nguyen) [reply ios/Runner/Info.plist:9]: the library [url_launcher](https://pub.dev/packages/url_launcher) has a method `canLaunch(url)` method which using `canOpenURL(_:)` in native iOS code. To use that method we have to declare supported scheme by `LSApplicationQueriesSchemes`
[https://developer.apple.com/documentation/uikit/uiapplication/1622952-canopenurl](https://developer.apple.com/documentation/uikit/uiapplication/1622952-canopenurl)

File: lib/presentation/widget/enter_otp/enter_otp_viewmodel.dart [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    return OnlineThunkAction((Store<AppState> store) async {
+      await _getPermanentTokenInteractor
+          .execute(_enterOTPArgument.baseUrl, _enterOTPArgument.email, _enterOTPArgument.password, otpCode: OTPCode(otp))
+          .then(
+              (result) => result.fold(
+                  (failure) => {
+                    // todo show toast message here
+                    _appToast.showErrorToast('invalid otp')
+                  },
+              ((success) => _loginSuccess())));
```
Comment (dab246) [inline lib/presentation/widget/enter_otp/enter_otp_viewmodel.dart:81]: I think indentation should be done here.

File: data/lib/src/datasource/authentication_datasource.dart [hunk 4]
```diff
@@ -65,6 +69,9 @@ class AuthenticationDataSource {
     });
   }
 
+  bool isAuthenticateWithOTPError(String authErrorCode) =>
+      authErrorCode != null && (authErrorCode == '1002' || authErrorCode == '1003');
```
Comment (dab246) [inline data/lib/src/datasource/authentication_datasource.dart:73]: I think it's a good idea to create an error list for easy testing and expansion later. `['1002', '1003']`

File: lib/presentation/widget/second_factor_authentication/second_factor_authentication_widget.dart [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+class SecondFactorAuthenticationWidget extends StatelessWidget {
+  final imagePath = getIt<AppImagePaths>();
+  final viewModel = getIt<SecondFactorAuthenticationViewModel>();
+
+  @override
+  Widget build(BuildContext context) {
+    final SecondFactorAuthenticationArguments arguments = ModalRoute.of(context).settings.arguments;
+    viewModel.setSecondFactorAuthenticationArguments(arguments);
+
+    return Container(
```
Comment (dab246) [inline lib/presentation/widget/second_factor_authentication/second_factor_authentication_widget.dart:50]: I find it a bit long. You can write it down as a separate function for readability. WDYT.

**Review by dab246 — APPROVED** [review]
LGTM

File: data/lib/src/datasource/authentication_datasource.dart (hunk 4 above)
Comment (manhduc-nguyen) [reply data/lib/src/datasource/authentication_datasource.dart:73]: hmm, error code list for only OTP case, I don't think we need that. If we have errorCode for other cases than OTP error then we have to use another list. this is not generic enough IMO.

File: data/lib/src/datasource/authentication_datasource.dart (hunk 1 above)
Comment (manhduc-nguyen) [reply data/lib/src/datasource/authentication_datasource.dart]: done

File: lib/presentation/widget/enter_otp/enter_otp_viewmodel.dart (hunk 3 above)
Comment (manhduc-nguyen) [reply lib/presentation/widget/enter_otp/enter_otp_viewmodel.dart]: done

File: lib/presentation/widget/second_factor_authentication/second_factor_authentication_widget.dart (hunk 5 above)
Comment (manhduc-nguyen) [reply lib/presentation/widget/second_factor_authentication/second_factor_authentication_widget.dart]: done

File: data/lib/src/datasource/authentication_datasource.dart (hunk 4 above)
Comment (hoangdat) [reply data/lib/src/datasource/authentication_datasource.dart:74]: it is so magical, please use constant instead

File: data/lib/src/network/linshare_http_client.dart [hunk 6]
```diff
@@ -77,11 +77,15 @@ class LinShareHttpClient {
...
       String password,
-      PermanentTokenBodyRequest bodyRequest) async {
+      PermanentTokenBodyRequest bodyRequest,
+      {OTPCode otpCode}) async {
     final basicAuth = 'Basic ' + base64Encode(utf8.encode('$userName:$password'));
 
     final headerParam = _dioClient.getHeaders();
     headerParam[HttpHeaders.authorizationHeader] = basicAuth;
+    if (otpCode != null && otpCode.value.isNotEmpty) {
+      headerParam['x-linShare-2fa-pin'] = otpCode.value;
```
Comment (hoangdat) [inline data/lib/src/network/linshare_http_client.dart:87]: no need a constant too

**Review by huynguyennovem — COMMENTED** [review]
/lgtm

File: data/lib/src/datasource/authentication_datasource.dart (hunk 4 above)
Comment (manhduc-nguyen) [reply data/lib/src/datasource/authentication_datasource.dart]: done

**Review by hoangdat — APPROVED** [review]

**Pull request merged**